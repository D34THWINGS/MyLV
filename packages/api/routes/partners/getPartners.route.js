const Joi = require('joi')
const moment = require('moment')
const qs = require('qs')
const config = require('@cracra/config/server')

const hasRole = require('../../helpers/hasRole.pre')
const getWorkingDays = require('../../helpers/getWorkingDays.helper')

const compensatedMealLabels = ['Production', 'Conferences']

function getLunchesForPartner(lunches, partner) {
  return lunches.filter(lunch =>
    lunch.owner.toString() === partner.id || lunch.attendees.some(attendeeId => attendeeId.toString() === partner.id))
}

module.exports = {
  method: 'GET',
  path: '/api/partners',
  config: {
    validate: {
      query: {
        page: Joi.number().min(1),
        limit: Joi.number().min(1).max(100),
        search: Joi.string().max(30),
        date: Joi.string().regex(/^\d{4}-\d{2}/),
      },
    },
    pre: [hasRole(config.cracra.partnersRoles)],
  },
  async handler(request, reply) {
    const query = qs.stringify(request.query)
    const date = request.query.date || moment().format('YYYY-MM')

    const res = await request.app.lvConnect.api(`/users?${query}`)
    const data = await res.json()
    const [entries, lunches] = await Promise.all([
      request.server.app.models.Entry.find({
        userId: { $in: data.results.map(partner => partner.id) },
        date: { $regex: date },
      }),
      request.server.app.models.Lunch.find({
        date: { $gt: moment(date, 'YYYY-MM').toDate() },
      }),
    ])

    const expectedDays = getWorkingDays(date)

    reply.mongodb(Object.assign({}, data, {
      results: data.results.map((partner) => {
        const partnerEntries = entries.filter(entry => entry.userId.toString() === partner.id)
        const mealVouchers = partnerEntries.reduce((nb, entry) =>
          (/-am$/.test(entry.date) && compensatedMealLabels.indexOf(entry.label) >= 0 ? nb + 1 : nb), 0)
        const lunchesCount = getLunchesForPartner(lunches, partner).length
        return Object.assign({}, partner, {
          mealVouchers: Math.max(mealVouchers - lunchesCount, 0),
          entryCounts: partnerEntries.reduce((acc, entry) => Object.assign(acc, {
            [entry.label]: (acc[entry.label] || 0) + 0.5,
          }), {}),
          lunchesCount,
          isWorklogComplete: partnerEntries.length / 2 >= expectedDays,
        })
      }),
    }))
  },
}
