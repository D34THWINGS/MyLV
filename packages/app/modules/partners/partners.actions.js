import * as qs from 'qs'

import { lvConnect } from '../auth/lvconnect'
import { fetchWithAuth } from '../auth/auth.actions'

export const PARTNERS_FETCH_SEARCH_SUCCESS = 'PARTNERS_FETCH_SEARCH_SUCCESS'
export const fetchPartnersSearch = (search, excludeSelf) => (dispatch, getState) =>
  lvConnect.api(`/users?search=${search}&limit=5`)
    .then(res => res.json())
    .then(data => dispatch({
      type: PARTNERS_FETCH_SEARCH_SUCCESS,
      payload: !excludeSelf ? data : {
        ...data,
        results: data.results.filter(partner => partner.id !== getState().auth.user.id),
      },
    }))
    .catch(e => console.error(e))

export const PARTNERS_CLEAR_SEARCH = 'PARTNERS_CLEAR_SEARCH'
export const clearPartnersSearch = () => ({ type: PARTNERS_CLEAR_SEARCH })

export const PARTNERS_FETCH_SUCCESS = 'PARTNERS_FETCH_SUCCESS'
export const fetchPartnersSuccess = data => ({ type: PARTNERS_FETCH_SUCCESS, payload: data })

export const PARTNERS_FETCH_START = 'PARTNERS_FETCH_START'
export const PARTNERS_FETCH_ERROR = 'PARTNERS_FETCH_ERROR'
export const fetchPartners = (params = { page: 1 }) => (dispatch) => {
  dispatch({ type: PARTNERS_FETCH_START, payload: params })

  const query = qs.stringify(params)

  return dispatch(fetchWithAuth(`/api/partners?${query}`))
    .then(data => dispatch(fetchPartnersSuccess(data)))
    .catch(() => dispatch({ type: PARTNERS_FETCH_ERROR }))
}

export const notifyAllPartners = () => dispatch => dispatch(fetchWithAuth('/api/worklog/notify', { method: 'POST' }))
