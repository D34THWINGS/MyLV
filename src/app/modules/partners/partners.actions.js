import { lvConnect } from '../auth/lvconnect'

export const PARTNERS_FETCH_SEARCH_SUCCESS = 'PARTNERS_FETCH_SEARCH_SUCCESS'
export const fetchPartnersSearch = (search, excludeSelf) => (dispatch, getState) =>
  lvConnect.api(`/users?search=${search}`)
    .then(res => res.json())
    .then(data => dispatch({
      type: PARTNERS_FETCH_SEARCH_SUCCESS,
      payload: !excludeSelf ? data : {
        ...data,
        results: data.results.filter(partner => partner.id !== getState().auth.user.id)
      }
    }))

export const PARTNERS_CLEAR_SEARCH = 'PARTNERS_CLEAR_SEARCH'
export const clearPartnersSearch = () => ({ type: PARTNERS_CLEAR_SEARCH })

export const PARTNERS_FETCH_SUCCESS = 'PARTNERS_FETCH_SUCCESS'
export const fetchPartnersSuccess = data => ({ type: PARTNERS_FETCH_SUCCESS, payload: data })
