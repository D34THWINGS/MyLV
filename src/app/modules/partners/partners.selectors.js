import { createSelector } from 'reselect'

const getPartnersById = state => state.partners.partnersById

export const getPartnersSearchResults = createSelector(
  getPartnersById,
  state => state.partners.partnersSearchResults,
  (partnersById, partnersSearchResults) => partnersSearchResults.map(partnerId => partnersById[partnerId]),
)

export const getPartnersList = createSelector(
  getPartnersById,
  state => state.partners.partnersList,
  (partnersById, partnersList) => partnersList.map(partnerId => partnersById[partnerId]),
)
