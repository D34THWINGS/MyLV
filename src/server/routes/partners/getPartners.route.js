const Joi = require('joi')
const moment = require('moment')
const qs = require('qs')
const config = require('config')

const lvConnect = require('../../helpers/lvconnect.helper')
const hasRole = require('../../helpers/hasRole.pre')

const compensatedMealLabels = ['Production', 'Conferences']

module.exports = {
  method: 'GET',
  path: '/api/partners',
  config: {
    validate: {
      query: {
        page: Joi.number().min(1),
        limit: Joi.number().min(1).max(100),
        search: Joi.string().max(30)
      }
    },
    pre: [hasRole(config.cracra.partnersRoles)]
  },
  async handler (request, reply) {
    const query = qs.stringify(request.query)

    const res = await lvConnect.api(`/users?${query}`)
    const data = await res.json()
    const [entries, lunches] = await Promise.all([
      request.server.app.models.Entry.find({
        userId: { $in: data.results.map(partner => partner.id) },
        date: { $regex: moment().format('YYYY-MM-') }
      }),
      request.server.app.models.Lunch.find({
        $or: [
          { attendees: request.auth.credentials.id },
          { owner: request.auth.credentials.id }
        ],
        date: { $gt: moment(`${moment().year()}-${moment().month()}`, 'YYYY-MM').toDate() }
      })
    ])

    reply.mongodb(Object.assign({}, data, {
      results: data.results.map(partner => {
        const partnerEntries = entries.filter(entry => entry.userId.toString() === partner.id)
        const mealVouchers = partnerEntries.reduce((nb, entry) =>
          /-am$/.test(entry.date) && compensatedMealLabels.indexOf(entry.label) >= 0 ? nb + 1 : nb, 0)
        return Object.assign({}, partner, {
          mealVouchers: Math.max(mealVouchers - lunches.length, 0),
          entryCounts: partnerEntries.reduce((acc, entry) => Object.assign(acc, {
            [entry.label]: (acc[entry.label] || 0) + 0.5
          }), {}),
          lunchesCount: lunches.length
        })
      })
    }))
  }
}
