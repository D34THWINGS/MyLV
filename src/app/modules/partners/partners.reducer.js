import { PARTNERS_CLEAR_SEARCH, PARTNERS_FETCH_SEARCH_SUCCESS, PARTNERS_FETCH_SUCCESS } from './partners.actions'

const initialState = {
  partnersById: {},
  partnersSearchResults: []
}

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case PARTNERS_FETCH_SEARCH_SUCCESS:
      return {
        ...state,
        partnersById: {
          ...state.partnersById,
          ...payload.results.reduce((acc, partner) => ({ ...acc, [partner.id]: partner }), {})
        },
        partnersSearchResults: payload.results.map(partner => partner.id)
      }
    case PARTNERS_CLEAR_SEARCH:
      return {
        ...state,
        partnersSearchResults: []
      }
    case PARTNERS_FETCH_SUCCESS:
      return {
        ...state,
        partnersById: {
          ...state.partnersById,
          ...payload.results.reduce((acc, partner) => ({ ...acc, [partner.id]: partner }), {})
        }
      }
    default:
      return state
  }
}
