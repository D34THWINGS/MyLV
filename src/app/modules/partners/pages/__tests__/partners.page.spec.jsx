import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import { PartnersPage } from '../partners.page'

jest.unmock('../partners.page')

describe('PartnersPage', () => {
  let props

  const getWrapper = () => shallow(<PartnersPage {...props} />)

  beforeEach(() => {
    props = {
      fetchPartners: jest.fn(),
      push: jest.fn(),
      partners: [{
        id: 'my_anaconda_dont',
        firstName: 'hello',
        lastName: 'world',
        entryCounts: {
          foo: 10,
        },
      }],
      isLoading: false,
      labels: {
        foo: 'bar',
        baz: 'qux',
      },
      classes: {},
      pageCount: 10,
      limit: 25,
      match: {
        params: {
          page: '1',
        },
      },
    }
  })

  it('should render partners list', () => {
    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render loader when fetching', () => {
    // Given
    props.isLoading = true

    // When
    const wrapper = getWrapper()

    // Then
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
