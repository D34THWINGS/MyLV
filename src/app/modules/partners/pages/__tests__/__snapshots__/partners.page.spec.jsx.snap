// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartnersPage should render loader when fetching 1`] = `<WithStyles(LoadingPage) />`;

exports[`PartnersPage should render partners list 1`] = `
<WithStyles(Paper)>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Partners | 
      FooApp
    </title>
  </HelmetWrapper>
  <WithStyles(Toolbar)>
    <WithStyles(Typography)
      component="h2"
      gutterBottom={true}
      variant="headline"
    >
      Partners
    </WithStyles(Typography)>
  </WithStyles(Toolbar)>
  <div>
    <WithStyles(Table)>
      <WithStyles(TableHead)>
        <WithStyles(TableRow)>
          <WithStyles(TableCell)>
            Nom
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Production"
            numeric={true}
          >
            Production
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Congés payés"
            numeric={true}
          >
            Congés payés
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Congés sans solde"
            numeric={true}
          >
            Congés sans solde
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Absences Syntec"
            numeric={true}
          >
            Absences Syntec
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Contributions"
            numeric={true}
          >
            Contributions
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Conférences"
            numeric={true}
          >
            Conférences
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="Arrêt Maladie"
            numeric={true}
          >
            Arrêt Maladie
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            Déjeuners
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            Ticket restaurants
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableHead)>
      <WithStyles(TableBody)>
        <WithStyles(TableRow)
          classes={
            Object {
              "root": "incompleteWorklog",
            }
          }
          hover={true}
          key="my_anaconda_dont"
        >
          <WithStyles(TableCell)>
            hello
             
            world
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="production"
            numeric={true}
          >
            9
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="paidHolidays"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="unpaidHolidays"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="conventionalHolidays"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="contributions"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="conferences"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="sickness"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            10
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            21
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          hover={true}
          key="booyah"
        >
          <WithStyles(TableCell)>
            foo
             
            bar
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="production"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="paidHolidays"
            numeric={true}
          >
            10
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="unpaidHolidays"
            numeric={true}
          >
            42
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="conventionalHolidays"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="contributions"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="conferences"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="sickness"
            numeric={true}
          >
            0
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            42
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            19
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </div>
  <table>
    <TableFooter
      component="tfoot"
    >
      <WithStyles(TableRow)>
        <WithStyles(TablePagination)
          count={250}
          labelDisplayedRows={[Function]}
          onChangePage={[Function]}
          onChangeRowsPerPage={[Function]}
          page={0}
          rowsPerPage={25}
          rowsPerPageOptions={
            Array [
              25,
              50,
              100,
            ]
          }
        />
      </WithStyles(TableRow)>
    </TableFooter>
  </table>
  <WithStyles(CardActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[MockFunction]}
    >
      Envoyer un rappel
    </WithStyles(Button)>
  </WithStyles(CardActions)>
</WithStyles(Paper)>
`;
